Cookbook/Proposals/Danger-Integration.md

```diff
@@ -0,0 +1,195 @@
+
+---
+
+On a side note, it's easy enough to test your Dangerfile for errors, by locally running in your terminal either `bundle exec danger pr <url-of-pr-to-test>` for a Ruby Dangerfile or `swift run danger-swift pr <url-of-pr-to-test>` for a Swift Dangerfile. So avoiding syntax errors and doing iteration testing bring similar experience in both worlds.
+
+In the end it's mainly a matter of tradeoff, the familiarity of the Swift language vs the additional cache step, dependencies, and the risk of plugins and documentation shortage (compare the [Danger-Swift documentation](https://danger.systems/swift/) and [DSL reference](https://danger.systems/swift/reference.html) to its more mature [Ruby counterpart](https://danger.systems/ruby/) and [DSL reference](https://danger.systems/reference.html) to get an idea of each project's doc maturity)
+
+### Integration in the CI
+
+We could potentially trigger the `danger` CLI earlier in the CI workflow, to be notified about anything we missed ASAP in the PR. But this brings the risk of not having access to the code (if we do that before the checkout step), requiring us to fetch the `Dangerfile` via the GitHub API in order for it to be present to run `danger`, and limiting our possibilties at what rule we can write in the future (and to be honest I'm not even sure every feature of Danger would work if we haven't checked out the code yet)
+
+Another alternative to trigger it earlier would be to run it during the `checkout_code_steps`, right after the `save_gems_cache` step, to run Danger right after its installation by `bundle install` in this job. This would allow to avoid waiting for things like the CocoaPods cache restoration and the `persist_workspace` steps, which could take a significant time to run. One drawback of this approach is that it means that `danger` would be run for every workflow using this `checkout_code` job (so basically all workflows, including the ones where no PR is associated, making `danger` fail on those), instead of only triggering on `test_pr` and `test_release` workflows.
+
+We could also decide to trigger `danger` later, for example _after_ the `test_babylon` job. This would give the advantage of allowing Danger to have access to test results for example. Which means that in the future we could use Danger to post a comment in the PR to tell which test failed or post the link to the HTML report maybe. But following that logic we should also trigger `danger` after `test_telus` and other jobs running tests, which will likely report some test failures multiple times (one per job if the failure is common to those targets). Which brings other questions and complexities into consideration (like does that mean one Danger comment on the PR... per test job, in each PR? Or a dedicated job after everything synchronizing all the previous test jobs but having to wait on them all before being able to report?) that might be too early to discuss for this initial proposal.
```

Comment:
> We could also decide to trigger `danger` later, for example _after_ the `test_babylon` job. This would give the advantage of allowing Danger to have access to test results for example. Which means that in the future we could use Danger to post a comment in the PR to tell which test failed or post the link to the HTML report maybe. 

I'm working on proposal for this. I think PR etiquette checks and build/test failure reporting could be done separately. This is related to issue that there's currently no way to wait for/hook into the end of the workflow (e.g. to post all those reports together) or to run extra job after failing one (e.g. fan-out to run test checks, fan-in to report any failures).

> Which brings other questions and complexities into consideration (...) that might be too early to discuss for this initial proposal.

There's a lot to discuss about on how to approach those reports, so we can take it to my proposal once I submit it.